core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -0,0 +1,249 @@
+    public String toString(String field) {
+        return "shard_splitting_query";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return sameClassAs(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return classHash();
+    }
+
+    private static void findSplitDocs(String idField, Predicate<BytesRef> includeInShard,
+                                      LeafReader leafReader, IntConsumer consumer) throws IOException {
+        Terms terms = leafReader.terms(idField);
+        TermsEnum iterator = terms.iterator();
+        BytesRef idTerm;
+        PostingsEnum postingsEnum = null;
+        while ((idTerm = iterator.next()) != null) {
+            if (includeInShard.test(idTerm) == false) {
+                postingsEnum = iterator.postings(postingsEnum);
+                int doc;
+                while ((doc = postingsEnum.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                    consumer.accept(doc);
+                }
+            }
+        }
+    }
+
+    private static final class Visitor extends StoredFieldVisitor {
+        int leftToVisit = 2;
+        final BytesRef spare = new BytesRef();
+        String routing;
+        String id;
+
+        void reset() {
+            routing = id = null;
+            leftToVisit = 2;
+        }
+
+        @Override
+        public void binaryField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            switch (fieldInfo.name) {
+                case IdFieldMapper.NAME:
+                    id = Uid.decodeId(value);
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected field: " + fieldInfo.name);
+            }
+        }
+
+        @Override
+        public void stringField(FieldInfo fieldInfo, byte[] value) throws IOException {
+            spare.bytes = value;
+            spare.offset = 0;
+            spare.length = value.length;
+            switch (fieldInfo.name) {
+                case RoutingFieldMapper.NAME:
+                    routing = spare.utf8ToString();
+                    break;
+                default:
+                    throw new IllegalStateException("Unexpected field: " + fieldInfo.name);
+            }
+        }
+
+        @Override
+        public Status needsField(FieldInfo fieldInfo) throws IOException {
+            switch (fieldInfo.name) {
+                case IdFieldMapper.NAME:
+                case RoutingFieldMapper.NAME:
+                    leftToVisit--;
+                    return Status.YES;
+                default:
+                    return leftToVisit == 0 ? Status.STOP : Status.NO;
+
+            }
+        }
+    }
+
+    /**
+     * This DISI visits every live doc and selects all docs that don't belong into this
+     * shard based on their id and rounting value. This is only used in a routing partitioned index.
+     */
+    private final class RoutingPartitionedDocIdSetIterator extends DocIdSetIterator {
```

Comment:
It is theoretically illegal to use live docs in a query since it prevents caching on the core key. However you could implement this more efficiently with a two-phase iterator, and the good news is that Lucene will make sure to check live docs before calling matches(). You could pass a `DocIdSetIterator.all(maxDoc)` as an approximation.

Reply:
good stuff I will explore this.